lib/AFS/CellCC/Util.pm

```diff
@@ -331,4 +332,58 @@ scratch_ok($$$$$) {
     return 1;
 }
 
+# Return the checksum of the file received as an argument.
+# The checksum returned by this function will be in following
+# format: "algorithm:checksum", where algorithm is the algorithm
+# of choice (e.g. MD5) and checksum is the checksum of the file
+# received as an argument. This checksum will be in hexadecimal form.
+sub
+calc_checksum($$$$$$) {
+    my ($dumpfh, $filesize, $algo, $jobid, $dvref, $state) = @_;
+
+    if ($filesize < 0) {
+        die("The size of the file cannot be negative\n");
+    }
+
+    my $start = time();
+    my $now = $start;
+
+    my $total = pretty_bytes($filesize);
+    my $bytes;
+    my $nbytes;
+
+    my $digest = Digest->new($algo);
+    my $checksum;
+
+    my $buf;
+    my $descr;
+
+    while (1) {
+        $nbytes = read($dumpfh, $buf, 16384);
+
+        if ($nbytes == 0) {
+            last;
+        }
+        if (!defined($nbytes)) {
+            die("Read of dump file failed: $!\n");
+        }
+        $digest->add($buf);
+
+        $now = time();
+        if ($now < $start || $now - $start > 60) {
+            $bytes = pretty_bytes(tell($dumpfh));
```

Comment:
Looking at this now, I'd rather have just a little more error checking. Something like:

my $pos = tell($dumpfh);
if ($pos < 0) {
    die("'tell' failed: $!\n");
}
$bytes = pretty_bytes($pos);